controller/src/main/java/io/pravega/controller/server/retention/AutoRetentionBucketService.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.retention;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractService;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.impl.StreamImpl;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.controller.store.stream.OperationContext;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.util.Config;
+import io.pravega.controller.util.RetryHelper;
+import lombok.extern.slf4j.Slf4j;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class AutoRetentionBucketService extends AbstractService implements BucketChangeListener {
+
+    private final int bucketId;
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final ScheduledExecutorService executor;
+    private final ConcurrentMap<Stream, CompletableFuture> retentionFutureMap;
+    private final LinkedBlockingQueue<BucketChangeListener.StreamNotification> notifications;
+    private final AtomicBoolean stop;
+    private CompletableFuture<Void> notificationLoop;
+
+    AutoRetentionBucketService(int bucketId, StreamMetadataStore streamMetadataStore,
+                               StreamMetadataTasks streamMetadataTasks, ScheduledExecutorService executor) {
+        this.bucketId = bucketId;
+        this.streamMetadataStore = streamMetadataStore;
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.executor = executor;
+
+        this.notifications = new LinkedBlockingQueue<>();
+        this.stop = new AtomicBoolean(false);
```

Comment:
https://github.com/pravega/pravega/pull/2046#discussion_r149852906

Yes. You can use `this::isRunning` as a stop condition in your Futures loop.

Check out `io.pravega.segmentstore.server.writer.StorageWriter`. It extends from an `AbstractThreadPoolService` which I wrote to be a base class for services such as yours. You don't have to extend from it, but if you look at the `doRun` method, it runs in a loop and uses `this::canRun` as a condition (which internally checks for `isRunning`).

Reply:
the problem is we use this before we notifyStarted. 
Which means the loop might find it not running and exit. 

So if we have to use isRunning instead of a separate atomicBoolean, then i can move some part of the code after notifyStarted()